enhancements/ingress/lb-subnet-selection-aws.md

```diff
@@ -0,0 +1,561 @@
+
+On the other hand, when the IngressController is created and the LoadBalancer-type
+service is not yet provisioned, these same invalid `Subnets` values will
+prevent the load balancer from being provisioned. The existing Ingress
+Operator logic will clearly indicate to the cluster admin that the load balancer
+failed to provision via the `LoadBalancerReady` status condition. In addition,
+the `LoadBalancerReady` condition will include the CCM event logs that indicate
+to the cluster admin that the `Subnets` value is invalid. Therefore, immutability
+provides a way for the Ingress Operator to produce a clear signal to the user that
+`Subnets` is invalid.
+
+Additionally, immutability prevents upgrade incompatibilities. If `Subnets`
+cannot be changed after the IngressController has been created, then we can
+leave unmanaged `service.beta.kubernetes.io/aws-load-balancer-subnets`
+service annotations as they are (i.e., don't remove or update). For more details
+on how this will be implemented see [Maintaining Upgrade Compatibility](#maintaining-upgrade-compatibility).
+
+If in the future a customer needs the `Subnets` field to mutable, this proposal
+can be updated to support that, as changing from immutable to mutable is a
+compatible API change.
+
+### Install Config Proposal (Day 1)
+
+TODO: Install time configuration proposal
+
+### Workflow Description
+
+#### Day 1 Workflow
+
+Setting the subnets for the default IngressController during installation:
+
+TODO: Workflow for install time configuration
+
+#### Day 2 Workflow
+
+Setting the subnets on an IngressController created after installation:
+
+1. A cluster admin, who wants to specify the subnets for a load balancer, creates an IngressController with
+  `spec.endpointPublishingStrategy.loadBalancer.providerParameters.aws.subnets` specified.
+2. The ingress operator reconciles the LoadBalancer Service and sets the
+   `service.beta.kubernetes.io/aws-load-balancer-subnets` annotation with the subnets provided by the user.
+
+### API Extensions
+
+This proposal will modify the `IngressController` API by adding a new
+field called `Subnets` of type `[]string` to the `AWSLoadBalancerParameters`
+struct type.
+
+TODO: What will install-time option modify?
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+This enhancement is directly enabling the https://issues.redhat.com/browse/XCMSTRAT-545
+feature for ROSA Hosted Control Planes.
+
+#### Standalone Clusters
+
+Standalone Clusters impact TBD
+
+#### Single-node Deployments or MicroShift
+
+This proposal does not have any special implications for single-node
+deployments or MicroShift.
+
+### Implementation Details/Notes/Constraints
+
+The Ingress Operator will use value of `spec.endpointPublishingStrategy.loadBalancer.providerParameters.aws.subnets`
+to set the `service.beta.kubernetes.io/aws-load-balancer-subnets` annotation
+on the LoadBalancer-type Service created by the ingress controller.
```

Comment:
@gcs278, can you add a note to call out explicitly that although this is consistent with what we did for `allowedSourceRanges`, this does constitute a divergence from earlier logic?  And note the reason we are doing it this way, namely that `status` should reflect the _actual_ status, and we do anticipate that the cluster-admin may have set the annotation manually (even though that isn't a supported operation).

Reply:
Yes sure. I talked a bit about `allowedSourceRanges` in the alternatives, but no problem with bringing it up here.

> namely that status should reflect the actual status

I'm confused here, do you mean spec? With `allowedSourceRanges`, the status should reflect the effective value on the Service. But the spec sometimes does not (when you remove `allowedSourceRanges` it doesn't clear `spec.loadBalancerSourceRanges`). Is that what you are trying to point out? I may be confused.

Reply:
Ah nvm, after our slack convo I think get it. The  https://github.com/openshift/enhancements/blob/master/enhancements/ingress/mutable-publishing-scope.md#as-a-cluster-administrator-i-want-my-default-ingresscontroller-to-be-private-internal proposal only set `status.endpointPublishingStrategy...` when the IC was created or service was recreated, which means it's not always reflecting the the effective value on the Service.

Reply:
Done.